### PR TITLE
Add a nacelle model to the c++ glue code

### PR DESCRIPTION
**Pull request by tonyinme**
Added a nacelle model to the cpp glue code.
This is meant to be used in Nalu.
The nacelle model uses a drag coefficient, area, and density to compute a force.
A Gaussian smoothing is used with the velocity sampled at the center.
A correction is added to the velocity sampled to take into account the slow-down of the flow.

### CODE REVIEW

**Comment by rafmudaf** [comment]
@sayerhs are you available to review this pull request?

File: glue-codes/fast-cpp/src/FAST_Prog.cpp [hunk 1]
```diff
@@ -22,7 +22,9 @@ void readTurbineData(int iTurb, fast::fastInputs & fi, YAML::Node turbNode) {
   }
   fi.globTurbineData[iTurb].numForcePtsBlade = turbNode["num_force_pts_blade"].as<int>();
   fi.globTurbineData[iTurb].numForcePtsTwr = turbNode["num_force_pts_tower"].as<int>();
-
+  fi.globTurbineData[iTurb].nacelle_cd = turbNode["nacelle_cd"].as<float>();
```
[the remark this answers is not in the document]
Comment (tonyinme) [reply glue-codes/fast-cpp/src/FAST_Prog.cpp]: Agreed, this has been added.

File: glue-codes/fast-cpp/src/OpenFAST.H [hunk 2]
```diff
@@ -30,6 +30,9 @@ struct globTurbineDataType {
   std::vector<double> TurbineHubPos;
   int numForcePtsBlade;
   int numForcePtsTwr;
+  float nacelle_cd;
```
[the remark this answers is not in the document]
Comment (tonyinme) [reply glue-codes/fast-cpp/src/OpenFAST.H]: Yes, this has been added.

File: glue-codes/fast-cpp/src/OpenFAST.H [hunk 3]
```diff
@@ -165,6 +171,17 @@ class OpenFAST {
   void stepNoWrite();
   void end();
 
+  // Compute the nacelle force
+  void calc_nacelle_force(float & u,
```
[the remark this answers is not in the document]
Comment (tonyinme) [reply glue-codes/fast-cpp/src/OpenFAST.H]: The const qualifier has been added to the corresponding variables.

File: glue-codes/fast-cpp/src/OpenFAST.cpp [hunk 4]
```diff
@@ -218,6 +218,26 @@ void fast::OpenFAST::step() {
      FAST_OpFM_Step(&iTurb, &ErrStat, ErrMsg);
      checkError(ErrStat, ErrMsg);
 
+     // Compute the force from the nacelle only if the drag coefficient is
+     //   greater than zero
+     if (nacelle_cd[iTurb]>0.) {
```
[the remark this answers is not in the document]
Comment (tonyinme) [reply glue-codes/fast-cpp/src/OpenFAST.cpp]: Yes, now, the new implementation makes it optional within openfast.

File: glue-codes/fast-cpp/src/OpenFAST.cpp [hunk 5]
```diff
@@ -287,6 +307,39 @@ void fast::OpenFAST::stepNoWrite() {
...
+        float & rho,
+        float & fx, 
+        float & fy, 
+        float & fz) {
+            // Calculate the force on the nacelle (fx,fy,fz) given the 
+            //   velocity sampled at the nacelle point (u,v,w), 
+            //   drag coefficient 'cd' and nacelle area 'area'
+    
+            // The constant pi
+            const float pi = acos(-1.0);
```
[the remark this answers is not in the document]
Comment (tonyinme) [reply glue-codes/fast-cpp/src/OpenFAST.cpp]: This has been fixed.

**Comment by tonyinme** [comment]
@sayerhs, all the changes have been made. Let me know if it is good now.

**Comment by rafmudaf** [comment]
@tonyinme is this pull request still on hold?

**Comment by tonyinme** [comment]
I'm still working on it, don't merge this yet.

**Comment by rafmudaf** [comment]
@tonyinme I heard this is ready to merge, can you confirm?

**Comment by tonyinme** [comment]
yes it is!

**Pull request merged**